### PR TITLE
Use `resize` instead of `append` to pad features

### PR DESCRIPTION
**Pull request by mlodato517**
**This Commit**
Updates various instances of `append(&mut vec![...])` with
`resize(...)`.

**Why?**
As a micro optimization. I don't expect this to affect any benchmarks
since the change will be so small compared to the time it takes the
model to do anything but I ran a small benchmark and this seemed to be
the fastest way to do this because (I think):

- we allocate each attention mask exactly once to the correct capacity
- we don't allocate a new vector to append to the existing one[^1]

And, while this won't speed up anything in practice, I think it might
read more clearly since `resize` tells you the final length so we can
see that all the vectors are the same final length and match `max_len`.

<details>
<summary>Micro benchmark run</summary>

The rust code was roughly:

```rust
pub fn append(len: usize, max: usize) -> Vec<usize> {
    let mut v = vec![1; len];
    v.append(&mut vec![0; max - v.len()]);
    v
}
pub fn resize(len: usize, max: usize) -> Vec<usize> {
    let mut v = Vec::with_capacity(max);
    v.resize(len, 1);
    v.resize(max, 0);
    v
}
pub fn halfway(len: usize, max: usize) -> Vec<usize> {
    let mut v = vec![1; len];
    v.resize(max, 0);
    v
}
pub fn overwrite(len: usize, max: usize) -> Vec<usize> {
    let mut v = vec![1; max];
    v[len..max].fill(0);
    v
}
```

and the parameters were roughly:

```rust
    for size in [10, 500, 1000, 5000] {
        for max in [size, size + 1, size * 2, size * 10] {
```

and `resize` was consistently the fastest. `halfway` was similar most of
the time but consistently slightly slower. `overwrite` was slower than
those for reasons I don't understand and `append` was consistently the
slowest (though, of course, the difference was very small when we were
appending zero or one elements).

</details>

[^1]: I can't really read assembly but in [this small godbolt
example][0] I see `__rust_alloc`, `__rust_alloc_zeroed`, and
`do_reserve_and_handle` so I don't think the compiler is seeing the
upcoming allocation and handling it all on the initial allocation.

[0]: https://godbolt.org/z/eTsnjn9Tq

### CODE REVIEW

File: src/pipelines/question_answering.rs [hunk 1]
```diff
@@ -903,6 +903,8 @@ impl QuestionAnsweringModel {
         spans
     }
 
+    /// Pads the passed features to the length of the longest feature. Masks are generated so the
```
Comment (mlodato517) [inline src/pipelines/question_answering.rs:906]: I totally guessed on this comment about the masks. I probably shouldn't have. Happy to remove/rework those!

File: src/pipelines/question_answering.rs [hunk 2]
```diff
@@ -914,23 +916,18 @@ impl QuestionAnsweringModel {
...
+                attention_mask.resize(max_len, 0);
                 attention_mask
             })
             .map(|input| Tensor::of_slice(&(input)))
             .collect::<Vec<_>>();
 
         for feature in features.iter_mut() {
-            feature
-                .offsets
-                .append(&mut vec![None; max_len - feature.input_ids.len()]);
```
Comment (mlodato517) [inline src/pipelines/question_answering.rs:927]: This is padding `feature.offsets` but is referencing `feature.input_ids`. I imagine the idea is that `feature.offsets.len() == feature.input_ids.len()` and we want everything to be `max_len` long. If that's not correct then whoops!

File: src/pipelines/question_answering.rs (hunk 1 above)
Comment (guillaume-be) [reply src/pipelines/question_answering.rs:906]: hi @mlodato517 - This is not entirely accurate, the masks are used to discard any answer generated that would be overlapping with these mask features. This ensure we cannot return as answer containing parts of the question (that is prepended to the context for QA) nor padding.

File: src/pipelines/question_answering.rs (hunk 2 above)
Comment (guillaume-be) [reply src/pipelines/question_answering.rs:927]: This is correct

File: src/pipelines/token_classification.rs [hunk 3]
```diff
@@ -950,26 +952,21 @@ impl TokenClassificationModel {
...
-            feature.input_ids.append(&mut vec![
-                self.tokenizer.get_pad_id().expect(
-                    "Only tokenizers with a padding index can be used for token classification"
-                );
-                max_len - feature.input_ids.len()
-            ]);
-            feature
-                .reference_feature
-                .append(&mut vec![false; max_len - feature.input_ids.len()]);
+            let padding_index = self.tokenizer.get_pad_id().expect(
```
Comment (guillaume-be) [inline src/pipelines/token_classification.rs:964]: This could be moved outside of the `for` loop

**Review by guillaume-be — COMMENTED** [review]
Thank you @mlodato517 that is indeed making the code lighter to read - not sure it has a lot of impact on performance. I have identified a few places where such code simplification can be done too - I will push them as part of this PR if that is OK with you

**Comment by guillaume-be** [comment]
Thank you for the PR!
